### ISSUE TITLE
Memory leak in Rust bindings

### ISSUE THREAD

**Issue by tazz4843** [alphacep/vosk-api#1036]
Hi there! I've been using Vosk in production, and I've noticed a major memory leak that causes severe issues. A bit of context: I've written a bot for the chat platform Discord, and it uses Vosk to run speech to text.
I create a Recognizer when a user begins speaking, and I destroy Recognizers after the user stops speaking and the result has been extracted from the Recognizer.

After roughly 45-60 minutes of audio being input, in small chunks (typically no longer than 10 seconds), the bot runs out of memory and dies.

I've recorded part of the leak with Heaptrack. It points to Vosk as a large part of the leak, with about 34.1MB leaked in Recognizer::FinalResult(), and 23.8MB leaked in Recognizer::AcceptWaveform() (side note, I'm feeding unsigned 16 bit int samples), and 1.3MB leaked in Recognizer::Recognizer() (constructor), over a time period of approximately 2 minutes.

Heaptrack also shows a massive 5.0GB leak in Model::Model(), despite my being sure to deallocate the model at the end of the program.
This Model leak is not as big of an issue as the slow climb to OOM, and if you were to ignore this one, I'd be perfectly okay with that.

Here's a screenshot of Heaptrack's bottom up view: 
![image](https://user-images.githubusercontent.com/60516372/174934918-db122146-41fe-4e22-9587-565c4509953d.png)
The top three items (leaking 2.5gb, 1.9gb, and 494.3MB, respectively) are sourced from the Model constructor, while the rest is from either the Model constructor, or from the various Recognizer functions.

The host is a VM with 16GB RAM on a Intel Xeon E5-2667 v4 CPU (2 socket host) with 16 cores allocated to the VM.

If you need more information, let me know and I'll try to get it to you.

**Comment by nshmyrev** [comment]
Model is not releasing data because you don't finalize the recognizers probably, its unreferenced by the recognizers.

What is your programming language? Are you doing it in pure C++?

**Comment by nshmyrev** [comment]
Feel free to reopen if you have additional information or need help.

**Comment by tazz4843** [comment]
Apologies for taking a bit long to respond.

I'm using Rust, with bindings to the C FFI, as here: https://github.com/Bear-03/vosk-rs.

I have been able to consistently reproduce this, no matter how I handle the Recognizers, whether that be freeing them as soon as I'm done with them, or holding on to them as long as possible.

**Comment by nshmyrev** [comment]
And what is the code to reproduce your issue then?

**Comment by tazz4843** [comment]
I solved this issue by using a different library, apologies for opening this issue.

**Comment by nshmyrev** [comment]
@tazz4843 which different library exactly?

**Comment by tazz4843** [comment]
Coqui: https://github.com/coqui-ai/STT

**Comment by chriskyndrid** [comment]
I believe I've encountered the same issue reported here. It's entirely possible I'm doing something incorrectly as well:

1) I create the Model:

```rust

pub static VOSK_MODEL: Lazy<Model> = Lazy::new(|| {
    let model_path = "./include/ai/vosk/model/gs";
    Model::new(model_path).expect("Failed to load vosk model.")
});
```

This calls (from the bindings)

```rust

  #[must_use]
  pub fn new(model_path: impl Into<String>) -> Option<Self> {
      let model_path_c = CString::new(model_path.into()).ok()?;
      let model_ptr = unsafe { vosk_model_new(model_path_c.as_ptr()) };

      Some(Self(NonNull::new(model_ptr)?))
  }

```

2) I receive audio buffers in the proper format from a gstreamer sink and accumulate those buffers. Once the buffers reach a certain point they follow a chain similar to:

Buffer -> Combine At Target Size -> Detect Start / End Position (VAD) -> Overlap Segments -> Denoise (neural network in parallel) -> Recognize Word(s) (in parallel) -> Deduplicate Words -> Add Punctuation (T5 Grammar Synthesis Model) -> Enhance -> Sentence Enhance (Another T5 Model) -> Sentence Encoder (MLM model). 

Data is pulled off and stored at various stages for various purposes. 

The ```Recognizer ``` is used as follows in parallel using Rayon:

```rust

let mut recognized_words = vad_chunks
            .par_iter()
            .flat_map(|vad_chunk| {
                let mut recognizer = Recognizer::new(&*VOSK_MODEL, 16000.0).unwrap();
                recognizer.set_max_alternatives(0);
                recognizer.set_words(true);

                for sample in vad_chunk.slice.chunks(4000) {
                    recognizer.accept_waveform(sample);
                }

                let result = recognizer.final_result().single();
                if let Some(crm) = result {
                    let words = crm.result;
                    let recognized_words: Vec<RecognizedWord> = words.into_iter().map(move |word| {
                        let start_time = vad_chunk.start_pts + ((word.start * 1_000_000_000.0) as u64);
                        let end_time = start_time + ((word.end - word.start) * 1_000_000_000.0) as u64;
                        log::debug!(
                                "Recognized start/end time for word {}: start={}, end={}, vad_chunk.start_pts={}",
                                word.word,
                                start_time,
                                end_time,
                                vad_chunk.start_pts
                             );
                        RecognizedWord {
                            media_id: media_id.clone(),
                            start_time,
                            end_time,
                            word: word.word.to_owned(),
                            confidence: word.conf,
                        }
                    }).collect();
                    recognized_words
                } else {
                    Vec::new()
                }
            })
            .collect::<Vec<_>>();
            
```
The bindings used for this are:

```rust

    pub fn new(model: &Model, sample_rate: f32) -> Option<Self> {
        let recognizer_ptr = unsafe { vosk_recognizer_new(model.0.as_ptr(), sample_rate) };
        Some(Self(NonNull::new(recognizer_ptr)?))
    }
    
     pub fn set_max_alternatives(&mut self, max_alternatives: u16) {
        unsafe { vosk_recognizer_set_max_alternatives(self.0.as_ptr(), max_alternatives as c_int) }
    }
    
     pub fn set_words(&mut self, enable: bool) {
        unsafe { vosk_recognizer_set_words(self.0.as_ptr(), if enable { 1 } else { 0 }) }
    }

    pub fn accept_waveform(&mut self, data: &[i16]) -> DecodingState {
        // vosk_recognizer_accept_waveform and vosk_recognizer_accept_waveform_f are meant
        // to be used by languages that do not have an i16 type (those functions also take PCM 16-bit audio,
        // but represented as an f32 or i8). Those extra functions aren't needed in rust so they
        // will not be wrapped

        let decoding_state = unsafe {
            vosk_recognizer_accept_waveform_s(self.0.as_ptr(), data.as_ptr(), data.len() as i32)
        };

        DecodingState::from_c_int(decoding_state)
    }
    
    
```

Additionally when Recognizer goes out of scope:

```rust

impl Drop for Recognizer {
    fn drop(&mut self) {
        unsafe { vosk_recognizer_free(self.0.as_ptr()) }
    }
}

```

Is automatically called. The ```Model``` is being reused for every run throughout the lifetime of the program. This works great, however, memory steadily climbs for every transcoding run, and the amount of memory that appears to leak is proportional to the duration of the transcription.

I like using Vosk due to it's performance. On my machine, using this approach, I can perform STT on an hour worth of audio including the denoise, vad, detection, deduplication parts in a handful of minutes making good use of all available cores. In contrast, although whisper is much more accurate, so far, it's slower. 

Heaptrack reports a scenario very similar to what the OP reported.

**Comment by nshmyrev** [comment]
You'd better report your problem in separate issue, I don't see how it is related to Android.

I can take a look on this but to speedup the process you could create a small reproducible example of the issue, like if it leaks if you simply create recognizers in the loop.

We don't see serious leaks in Python for sure.

**Comment by chriskyndrid** [comment]
@nshmyrev , Thank you for taking the time to respond I appreciate and your work on this project. 

Do you use a separate core in your Android platform? In my research it appeared that there were a number of memory leaks reported that remained unresolved on various platforms, and I suspected it was perhaps related to the same fundamental issue(but I of course may be wrong). In any case, do you want a separate issue? This issue, although marked closed, is entirely relevant as the OP's heaptrack posted is representative of what I'm seeing. If you still want a separate issue, I can certainly create one.

The example I posted above illustrates the sum total of the interactions I'm performing with Vosk in rust, and I posted the ffi bindings the Rust program is interfacing with:

1) Model initialization

```rust
pub static VOSK_MODEL: Lazy<Model> = Lazy::new(|| {
    let model_path = "./include/ai/vosk/model/gs";
    Model::new(model_path).expect("Failed to load vosk model.")
});

```

2) Using a parallel iterator via the Rayon crate, on many threads based on chunking the input audio I:

```rust
let mut recognized_words = vad_chunks
            .par_iter()
            .flat_map(|vad_chunk| {
                let mut recognizer = Recognizer::new(&*VOSK_MODEL, 16000.0).unwrap();
                recognizer.set_max_alternatives(0);
                recognizer.set_words(true);

                for sample in vad_chunk.slice.chunks(4000) {
                    recognizer.accept_waveform(sample);
                }

                let result = recognizer.final_result().single();
               
                Vec::new()
            })
            .collect::<Vec<_>>();
```

[This](https://github.com/Bear-03/vosk-rs) is the rust bindings crate I'm using. I can create a sample program in Rust that performs these steps if it will help speed up the research process.

Thanks again for your time.
